pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -1340,18 +1341,42 @@ private void asyncDeleteCursorWithCleanCompactionLedger(PersistentSubscription s
             return;
         }
 
-        currentCompaction.handle((__, e) -> {
-            if (e != null) {
-                log.warn("[{}][{}] Last compaction task failed", topic, subscriptionName);
+        // Avoid concurrently execute compaction and unsubscribing.
+        synchronized (this) {
```

Comment:
Is there really a need for a synchronized block here?

Reply:
Yes, it is. It prevents the modifying of the variable `currentCompaction`.